File: packages/tactical-engine-py/tactical_engine/validation/action_idempotency.py
import time

DEFAULT_TTL_SECONDS = 10 * 60  # 10 minutes


class ActionIdempotencyTracker:
    def __init__(self, ttl_seconds: float = DEFAULT_TTL_SECONDS) -> None:
        self._seen: dict[str, float] = {}  # action_id -> timestamp
        self._ttl = ttl_seconds

    def has(self, action_id: str) -> bool:
        self._evict()
        return action_id in self._seen

    def record(self, action_id: str) -> None:
        self._evict()
        self._seen[action_id] = time.monotonic()

    def _evict(self) -> None:
        cutoff = time.monotonic() - self._ttl
        self._seen = {k: v for k, v in self._seen.items() if v >= cutoff}

File: packages/tactical-engine/src/validation/action-idempotency.ts
const DEFAULT_TTL_MS = 10 * 60 * 1000; // 10 minutes

export class ActionIdempotencyTracker {
  private readonly seen = new Map<string, number>(); // actionId -> timestamp
  private readonly ttlMs: number;

  constructor(ttlMs = DEFAULT_TTL_MS) {
    this.ttlMs = ttlMs;
  }

  has(actionId: string): boolean {
    this.evict();
    return this.seen.has(actionId);
  }

  record(actionId: string): void {
    this.evict();
    this.seen.set(actionId, Date.now());
  }

  private evict(): void {
    const cutoff = Date.now() - this.ttlMs;
    for (const [id, ts] of this.seen) {
      if (ts < cutoff) {
        this.seen.delete(id);
      }
    }
  }
}
